### PR TITLE
GNOME PAYG 3.32.0 Rebase

### PR DESCRIPTION
**Pull request by mazen-asef**
This is the Gnome 3.32.0 Pay As you Go Rebase.
I used small fixes as a squashing policy, no commit was dropped.

Picked:
48c673855 payg: Initial implementation of the PAYG manager component
8336d9e99 loginDialog: Fix vertical misalignment of items in the users list
0af9fb5d3 payg: Autostart the eos-payg daemon when constructing the PaygManager
2201e901e payg: Add an isLocked attribute to the PaygManager
8d17c7129 payg: Handle D-Bus errors from com.endlessm.Payg1
7822129cd payg: Add a new screen to allow unlocking PAYG machines
bb95e8f75 st-private: Add support for letter-spacing in ClutterText based StWidgets
f5f55a395 st-entry: Set text-related CSS properties on the internal ClutterText
419d4b858 payg: Visual changes to the code entry widget to make it stand out
e4a1b6601 payg: Integrate the unlocking dialog for PAYG in the screen shield
4f815dd88 payg: Allow querying the PaygManager whether it's initialized
1d12a7393 payg: Wait for the PaygManager before showing the screenshield's dialog
6ccc3ff9b payg: Make sure to initialize the users list after startup has completed
740dcf60d payg: Integrate the PAYG unlock screen with the login screen
e944815b6 payg: Show an explicit error message when timing out
aa9bd5264 payg: Don't auto-lock out of user sessions based on PAYG status
39d996973 payg: Use the right strings for text elements in the unlock dialog
da406ca8e payg: Add extra strings and update style for text in the unlock dialog
69f30a106 payg: Support the new property com.endlessm.Payg1.RateLimitEndTime
690a4f299 payg: Check if proxy was created
51aa84d7c payg: Show periodic notifications with the remaining PAYG time
984e74f53 payg: Show a notification with the remaining PAYG time on startup
90ce0a7e4 customerSupport: Manage vendor-customer-support.ini from its own class
e96653000 payg: Support image-specific overrides for text in the unlock dialog
1a71d0c7e payg: Show image-specific instructions for PAYG, if available
11899c4de paygUnlockDialog: deduplicate validating keypresses
ad8318102 paygManager: support validating codes against CodeFormat
42e8fa3e2 paygUnlockDialog: delegate validation to paygManager
e00520de0 payg: tweak code entry field for longer codes
12d9ff60c paygManager: don't overflow JS integers
f0eda0fc8 paygManager: make timeRemainingSecs() public
00572fbed messageTray: Don't expire a notification if its focus grabber is focused
8c3d99673 js/ui/payg.js: PayGo UI utilities
ce7348694 payg: make widgets insensitive if code successfully added
6af8b5884 payg: factor out reminder notifications
0302b7cd9 status/paygo: new applet for PayGo account details and actions
73ba532cc status/paygo: add menu action to apply a code before PayGo time is expired
eac1c1d49 payg: Consistently rename PayGo to Pay As You Go
58341e178  config: Remove obsolete notifications
b91d750ed paygManager: Understand ExpiryTime in terms of CLOCK_BOOTTIME
71b6f949d payg: Use CLOCK_BOOTTIME for rate limiting
862a597a2 payg: Properly update time remaining every minute
2b0f6600a payg: Add success screen

Squashed:
47dfaed20 payg: Make the PaygManager a permanent component, not a loadable one
116c44170 paygUnlockDialog: fix English string
31998eae7 status/payg: Fix applet visibility
994f70156 payg: rename nextButton -> applyButton to better match usage
c0b68495e payg: Add missing import
2bb68ce8a payg: Missing shell import

https://phabricator.endlessm.com/T26318

### CODE REVIEW

File: js/gdm/loginDialog.js [hunk 1]
```diff
@@ -112,27 +112,16 @@ var UserListItem = class {
...
-                }
-
-                return GLib.SOURCE_CONTINUE;
-            });
-
-        GLib.Source.set_name_by_id(this._timedLoginTimeoutId, '[gnome-shell] this._timedLoginTimeoutId');
-
+        this._timedLoginIndicator.add_style_pseudo_class('showing')
+        Tweener.addTween(this._timedLoginIndicator,
+                         { scale_x: 1.,
```
Comment (GeorgesStavracas) [inline js/gdm/loginDialog.js]: `1.0`

File: js/gdm/loginDialog.js [hunk 2]
```diff
@@ -112,27 +112,16 @@ var UserListItem = class {
...
+                           time: time,
+                           transition: 'linear',
+                           onComplete: function() {
```
Comment (GeorgesStavracas) [inline js/gdm/loginDialog.js]:
```javascript
onComplete: hold.release.bind(this),
```

File: js/gdm/loginDialog.js [hunk 3]
```diff
@@ -112,27 +112,16 @@ var UserListItem = class {
...
+                               hold.release();
+                           },
+                           onCompleteScope: this
```
Comment (GeorgesStavracas) [inline js/gdm/loginDialog.js]: You can remove this if the previous suggestion is applied.

File: js/gdm/loginDialog.js [hunk 4]
```diff
@@ -144,6 +133,7 @@ var UserListItem = class {
 
         this._timedLoginIndicator.visible = false;
         this._timedLoginIndicator.scale_x = 0.;
+        this._timedLoginIndicator.remove_style_pseudo_class('showing')
```
Comment (GeorgesStavracas) [inline js/gdm/loginDialog.js]: Missing `;` at the end of the line

File: js/misc/paygManager.js [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+        if (this._codeFormat === value)
+            return;
+
+        this._codeFormat = value;
+        try {
+            this._codeFormatRegex = new GLib.Regex(this._codeFormat,
+                                                   GLib.RegexCompileFlags.DOLLAR_ENDONLY,
+                                                   GLib.RegexMatchFlags.PARTIAL);
+        } catch (e) {
+            logError(e, `Error compiling CodeFormat regex "${this._codeFormat}"`);
```
Comment (GeorgesStavracas) [inline js/misc/paygManager.js]: This kind of string breaks gettext translations :(

Better not use it.

Comment (GeorgesStavracas) [reply js/misc/paygManager.js]: (But will be fixed at a very near future; the patch at https://savannah.gnu.org/bugs/?50920 landed)

File: js/ui/payg.js [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+    }
+
+    activate() {
+        // We get here if the Apply button is inactive when we try to click it.
+        // Unless we're already done, exit early so we don't destroy the
+        // notification)
+        if (this._verificationStatus != UnlockStatus.SUCCEEDED)
+            return;
+
+        this.parent();
```
Comment (GeorgesStavracas) [inline js/ui/payg.js]: `super.activate()`

File: js/ui/status/payg.js [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        GLib.Source.set_name_by_id(this._timeoutRefreshId, '[gnome-shell] this._timeoutRefresh');
+    }
+
+    _timeoutRefresh() {
+        this._updateTimeRemainingAndSyncWhenReady();
+        return GLib.SOURCE_CONTINUE;
+    }
+
+    _onDestroy() {
+        this.parent();
```
Comment (GeorgesStavracas) [inline js/ui/status/payg.js]: `super._onDestroy();`

File: js/misc/paygManager.js (hunk 5 above)
Comment (mazen-asef) [reply js/misc/paygManager.js]: But this is a log statement - would it be translated?

Comment (GeorgesStavracas) [reply js/misc/paygManager.js]: All the translations for **the entire file** will be broken. If there's any chance this file will receive a translatable string, now or in the future, this should be dropped.

File: js/misc/paygManager.js [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+        return this._initialized;
+    }
+
+    get enabled() {
+        return this._enabled;
+    }
+
+    get expiryTime() {
+        return this._expiryTime;
+
```
Comment (GeorgesStavracas) [inline js/misc/paygManager.js]: Nitpick: remove extra new line

File: js/ui/main.js [hunk 9]
```diff
@@ -234,8 +247,18 @@ function _initializeUI() {
     ExtensionSystem.init();
 
     if (sessionMode.isGreeter && screenShield) {
-        layoutManager.connect('startup-prepared', () => {
-            screenShield.showDialog();
+        layoutManager.connect('startup-prepared', function() {
```
Comment (GeorgesStavracas) [inline js/ui/main.js]: `() => {`

File: js/ui/payg.js [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+    _createApplyButton() {
+        let box = new St.BoxLayout();
+
+        let label = new St.Bin({ x_expand: true,
+                                 child: new St.Label ({
+                                                        x_expand: true,
+                                                        x_align: Clutter.ActorAlign.CENTER,
+                                                        text: _("Apply Code")
+                                                       })
+                               });
```
Comment (GeorgesStavracas) [inline js/ui/payg.js]: Nitpick: code style ↓

```javascript
let label = new St.Bin({ x_expand: true,
                         child: new St.Label ({ x_expand: true,
                                                x_align: Clutter.ActorAlign.CENTER,
                                                text: _("Apply Code") }) });
```

File: js/ui/payg.js [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+            Main.messageTray._expandActiveNotification();
+
+        this._notification.connect('destroy', () => {
+            this._notification = null;
+        });
+    }
+
+    clearNotification() {
+        if (this._notification)
+            this._notification.destroy();
```
Comment (GeorgesStavracas) [inline js/ui/payg.js]: Add `this._notification = null;` after destroying for extra safety

**Pull request merged**